rfcs/0000-wire-policies-with-backend-services.md

```diff
@@ -0,0 +1,188 @@
+
+* **Kuadrant supports only one instance deployed in a configurable namespace**
+
+Wiring is needed. The policy controllers do not know where Limitador/Authorino live. All policies will be linked to the same instance of Limitador/Authorino.
+
+* **Kuadrant supports multiple instances in a cluster**
+
+Wiring is needed. There should be a way to know the Limitador/Authorino instance only by reading the policy.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The proposal is to make Kuadrant to support multiple instances in a cluster.
+
+This proposal is based on a simple design decision: One gateway runs one and only one kuadrant's external auth service (Authorino) and only one kuadrant's external rate limit service (Limitador).
+
+Multiple rate limit or auth stages involving multiple instances of Authorino and Limitador can be implemented.
+Until there is a real use case and it is strictly necessary, this scenario is discarded for implementing kuadrant's protection.
+The main reason is about complexity. It is already complex enough to reason about rate limiting and auth services having a single stage. Adding multiple rate limiting stages, or hitting multiple Limitador instances in a single stage (doable with the WASM module) makes it too complex to reason about observed behavior. Currently there is no use case to require that complex scenario.
+
+A kuadrant instance includes:
+
+* One Limitador deployment instance
+* One Authorino deployment instance
+* A list of dedicated gateways. Those gateways cannot be shared between multiple kuadrant instances.
+
+A diagram to ilustrate some concepts:
+
+![](https://i.imgur.com/y7gQfRa.png)
+
+Highlights:
+* The Kuadrant instance is not enclosed by k8s namespaces.
+* One gateway can belong to (be managed by) one and only one kuadrant instance.
+* One kuadrant instance does not own rate limit policies or auth policies.
+* Each kuadrant instance owns one instance (possibly multiple replicas, though) of Limitador and one instance of Authorino. Those instances are shared among all gateways managed by the kuadrant instance.
+* The traffic routed by HTTPRoute 1 through the gateway A will be protected by RLP 1 and KAP 1, using Limitador and Authorino instances located at the namespace K1.
+* The traffic routed by HTTPRoute 2 through the gateway B will be protected by RLP 2 and KAP 2, using Limitador and Authorino instances located at the namespace K1.
+* The traffic routed by HTTPRoute 2 through the gateway C will be protected by RLP 2 and KAP 2, using Limitador and Authorino instances located at the namespace K2.
+* The HTTPRoute 2 example shows that when the traffic for the same service is routed through multiple gateways, at least for rate limiting, Kuadrant cannot keep consistent counters. The user would expect X rps and actually it would be X rps per gateway.
```

Comment:
IMO, this is the hardest limitation to sell? One who knows about the [global RL protocol](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/global_rate_limiting) may find counterintuitive that 2 gateways cannot share the same Limitador instance.

Could synchronization of the counters across multiple instances of Limitador mitigate this (as it makes fairly transparent that for 2 GWs there are actually 2 instances of Limitador)? Or perhaps it's better is the design allowed for a Kuadrant instance to take control of more than one GW (which seems possible with the implementation proposed based on label selectors)?

Reply:
I think it does not apply anymore. Check the new content.